<h1> Soal 1: Pemahaman Tentang Model Evaluasi</h1>

Jawab pertanyaan di bawah ini dengan bahasa masing-masing?

1. Apa perbedaan antara data latih, data validasi, dan data test?
2. Bagaimana cara kita menilai performa suatu model?
3. Apa itu Confusion Matrix? Jelaskan secara lengkap!
4. Apa itu Classification Report dari sklearn?

Jawab:

1. - Data latih adalah data yang digunakan untuk training model.
   - Data validasi digunakan untuk proses validasi dan mencegah overfitting.
   - Data test digunakan untuk testing model, sebagai simulasi penggunaan model.

2. Dengan cara menghitung nilai akurasi dari model menggunakan metrics accuracy yang sebelumnya telah dilakukan validasi

3. Confusion matrix juga sering disebut error matrix. Pada dasarnya confusion matrix memberikan informasi perbandingan hasil klasifikasi yang dilakukan oleh sistem (model) dengan hasil klasifikasi sebenarnya.
Terdapat 4 istilah sebagai representasi hasil proses klasifikasi pada confusion matrix. Keempat istilah tersebut adalah True Positive (TP), True Negative (TN), False Positive (FP) dan False Negative (FN).

4. Report dari klasifikasi yang telah kita lakukan, menampilkan 3 infromasi yaitu precision, recall dan f1-score berdasarkan confusion matrix

---

<h1>Soal 2: Aplikasi Model Evaluasi</h1>

Kali ini kita akan menggunakan data untuk memprediksi kelangsungan hidup pasien yang telah mengalami operasi payudara. Dengan informasi yang dimiliki terkait pasien, kita akan membuat model untuk memprediksi apakah pasien akan bertahan hidup dalam waktu lebih dari 5 tahun atau tidak.

Lebih Lengkapnya kalian bisa membaca informasi tentang dataset di link berikut: https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.names

Buat model Klasifikasi (Model/Algoritma Bebas) untuk memprediksi status pasien dengan ketentuan sebagai berikut:

1. Bagi kedua data ini menjadi data training dan data test dengan test_size=0.25.
3. Pelajar tentang metrics roc_auc_score kemudian buatlah model dan evaluasi dengan menggunakan teknik cross-validation dengan scoring 'roc_auc'. Baca https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html untuk menggunakan metric roc_auc saat cross-validation.
3. Berapa score rata2 dari model dengan teknik cross-validation tersebut?
4. Prediksi data test dengan model yang telah kalian buat!
5. Bagaimana hasil confusion matrix dari hasil prediksi tersebut?
6. Bagaimana classification report dari hasil prediksi tersebut?
5. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status positive?
6. Seberapa baik model anda dalam memprediksi seorang pasien mempunyai status negatif?

In [11]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [12]:
df.head()

Age  Patient's Years  N_positive_ax  survival_status
0   30               64              1                1
1   30               62              3                1
2   30               65              0                1
3   31               59              2                1
4   31               65              4                1

In [13]:
df['survival_status'].value_counts()

1    225
2     81
Name: survival_status, dtype: int64

In [14]:
X = df.drop('survival_status', axis=1)
y = df['survival_status']

In [15]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [16]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import cross_val_score

model_knn = KNeighborsClassifier()
model_knn.fit(X_train, y_train)

KNeighborsClassifier()

In [17]:
cvs = cross_val_score(model_knn, X_train, y_train, cv=5, scoring='roc_auc')
cvs

array([0.64166667, 0.80555556, 0.7375    , 0.69166667, 0.55714286])

In [18]:
cvs.mean()

0.6867063492063492

In [19]:
y_pred = model_knn.predict(X_test)
y_pred

array([1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1,
       1, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 2, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 2, 1, 2, 1, 2, 1, 1, 1], dtype=int64)

In [21]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, y_pred, labels = [1,2])

array([[42,  4],
       [27,  4]], dtype=int64)

In [23]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           1       0.61      0.91      0.73        46
           2       0.50      0.13      0.21        31

    accuracy                           0.60        77
   macro avg       0.55      0.52      0.47        77
weighted avg       0.56      0.60      0.52        77



In [ ]:
Model baik dalam memprediksi pasien positif dengan nilai f1-score 0.73

In [ ]:
Model tidak baik dalam memprediksi pasien negatif dengan nilai f1-score 0.21

---

<h1> Soal 3: Pemahaman Tentang Model Selection</h1>

Jelaskan dengan bahasa sendiri!

1. Apa itu Bias dan Variance?
2. Apa itu Overfitting dan Underfitting?
3. Apa yang bisa kita lakukan untuk mengatur kompleksitas dari model?
4. Bagaimana model yang baik?
5. Kapan kita menggunakan GridSearchcv dan kapan menggunakan RandomizedSearchCV?

Jawab:

1. - Bias adalah perbedaan antara rata rata hasil prediksi dari model Machine Learning yang kita develop dengan data nilai yang sebenarnya.
   - Variance adalah variabel dari prediksi model untuk data tertentu dimana memberikan kita informasi perserbaran data kita.

2. - Overfitting terjadi saat model menghapal semua data training, tanpa menganalisis pola trend dari data.
   - Underfitting adalah kebalikannya, model yang terlalu simple, dan tidak menganalisis data training secara keseluruhan.

3. Melakukan Tuning Hyperparameter

4. Model memiliki bias error yang kecil dan variance error yang kecil

5. Grid Search CV digunakan ketika hyperparameter yang digunakan kecil. Namun, jika jumlah parameter untuk dipertimbangkan sangat tinggi dan besarnya pengaruh tidak seimbang, pilihan yang lebih baik adalah menggunakan Randomized Search CV.

---

<h1> Soal 4: Aplikasi Model Selection</h1>

1. Bagi kedua data berikut ini menjadi data training dan data test dengan test_size=0.25.
2. Gunakan algoritma KNN sebagai model classifier.
3. Gunakan fungsi GridSearchCV untuk hyperparameter tuning dan model selection.
4. jumlah fold bebas!, gunakan scoring 'roc_auc'
5. Definisikan kombinasi hyperparameter untuk model selection dengan GridSearchCV. kombinasi Hyperparameter bebas, baca lagi dokumentasi KNN di link berikut https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html untuk memahami lagi jenis2 hyperparameter di algorithma KNN.
6. Latih model terhadap data training.
7. Apa hyperparameter terbaik untuk kombinasi hyperparameter kalian?
8. Berapa score validasi terbaik dari model tersebut?
9. Prediksi probabilitasi output dari model yang telah di buat terhadap data test. note : gunakan method .predict_proba() untuk menghasilkan output probabilitas
10. Perhatikan bahwa hasil prediksi ada 2, dimana masing2 adalah nilai probabilitas untuk setiap class label. Ambil nilai probabilitas pasien phositive meninggal dalam waktu kurang dari 5 tahun. note : gunakan bantuan attirubte .classes_ untuk mengetahui urutan label dari hasil prediksi probabilitas.
11. Berapa nilai score roc_auc untuk data test?
12. Apakah model anda termasuk baik, overtting, atau underfitting?
13. Ulangi tahap di atas namun kali ini menggunakan algoritma DecisionTreeClassifier dan kalian bisa menggunakan RandomizedSearchCV apabila process training lama. pelajari algoritma DecisionTreeClassifier di linkberikut: https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html?highlight=decisiontreeclassifier#sklearn.tree.DecisionTreeClassifier
14. Bandingkan scorenya dengan Algoritma KNN, mana yang lebih baik?

Note : Data Science adalah experiment, sangat di dimungkinkan memerlukan beberapa kali percobaan untuk mendapatkan hasil yang terbaik! Happy Coding :)

In [ ]:
import pandas as pd

url = 'https://raw.githubusercontent.com/jbrownlee/Datasets/master/haberman.csv'
list_cols = ['Age', "Patient's Years", "N_positive_ax", "survival_status"]
df = pd.read_csv(url, names=list_cols)

In [ ]:
df.head()

In [8]:
from sklearn.model_selection import GridSearchCV

param_grid = ({'n_neighbors':[1,2,3,4,5]})

gscv = GridSearchCV(model_knn, param_grid=param_grid, cv=5, scoring='roc_auc')

In [24]:
from sklearn.model_selection import GridSearchCV

param_grid = ({'n_neighbors':[1,2,3,4,5], 'weights':['uniform','distance']})

gscv = GridSearchCV(model_knn, param_grid=param_grid, cv=5, scoring='roc_auc')

In [25]:
gscv.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5],
                         'weights': ['uniform', 'distance']},
             scoring='roc_auc')

In [26]:
gscv.best_score_

0.6867063492063492

In [27]:
model_knn = gscv.best_estimator_

In [28]:
model_knn.predict_proba(X_test)

array([[0.8, 0.2],
       [0.6, 0.4],
       [0.8, 0.2],
       [1. , 0. ],
       [0.6, 0.4],
       [1. , 0. ],
       [1. , 0. ],
       [0.2, 0.8],
       [0.6, 0.4],
       [0.8, 0.2],
       [0.8, 0.2],
       [1. , 0. ],
       [0.6, 0.4],
       [0.6, 0.4],
       [0.6, 0.4],
       [0.8, 0.2],
       [0.4, 0.6],
       [1. , 0. ],
       [1. , 0. ],
       [1. , 0. ],
       [1. , 0. ],
       [0.6, 0.4],
       [0.8, 0.2],
       [1. , 0. ],
       [0.2, 0.8],
       [0.8, 0.2],
       [1. , 0. ],
       [0.6, 0.4],
       [0.8, 0.2],
       [0.8, 0.2],
       [1. , 0. ],
       [1. , 0. ],
       [0.6, 0.4],
       [0.6, 0.4],
       [0.8, 0.2],
       [1. , 0. ],
       [1. , 0. ],
       [0.6, 0.4],
       [0.4, 0.6],
       [1. , 0. ],
       [0.2, 0.8],
       [1. , 0. ],
       [0.6, 0.4],
       [1. , 0. ],
       [0.8, 0.2],
       [0.8, 0.2],
       [0.8, 0.2],
       [0.8, 0.2],
       [1. , 0. ],
       [1. , 0. ],
       [0.8, 0.2],
       [0.8, 0.2],
       [1. ,

In [29]:
model_knn.classes_

array([1, 2], dtype=int64)

In [31]:
y_pred = model_knn.predict_proba(X_test)[:, 1]
y_pred

array([0.2, 0.4, 0.2, 0. , 0.4, 0. , 0. , 0.8, 0.4, 0.2, 0.2, 0. , 0.4,
       0.4, 0.4, 0.2, 0.6, 0. , 0. , 0. , 0. , 0.4, 0.2, 0. , 0.8, 0.2,
       0. , 0.4, 0.2, 0.2, 0. , 0. , 0.4, 0.4, 0.2, 0. , 0. , 0.4, 0.6,
       0. , 0.8, 0. , 0.4, 0. , 0.2, 0.2, 0.2, 0.2, 0. , 0. , 0.2, 0.2,
       0. , 0.2, 0. , 0.2, 0. , 0.2, 0. , 0.2, 0.2, 0. , 0. , 0.4, 0.4,
       0. , 0. , 0. , 0.2, 0.6, 0. , 0.6, 0. , 0.6, 0. , 0.4, 0. ])

In [32]:
roc_auc_score(y_test, y_pred)

0.6297335203366059

In [33]:
from sklearn.tree import DecisionTreeClassifier

model_dtc = DecisionTreeClassifier()
model_dtc.fit(X_train, y_train)

DecisionTreeClassifier()

In [37]:
tree_param = {'criterion':['gini','entropy'],'max_depth':[4,5,6,7,8,9,10,11,12,15,20,30,40,50,70,90,120,150]}

gscv = GridSearchCV(model_dtc, param_grid=tree_param, cv=5, scoring='roc_auc')

In [38]:
gscv.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=DecisionTreeClassifier(),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [4, 5, 6, 7, 8, 9, 10, 11, 12, 15, 20, 30,
                                       40, 50, 70, 90, 120, 150]},
             scoring='roc_auc')

In [39]:
gscv.best_score_

0.628015873015873

In [40]:
model_dtc = gscv.best_estimator_

In [41]:
model_dtc.predict_proba(X_test)

array([[0.88333333, 0.11666667],
       [0.52173913, 0.47826087],
       [0.74074074, 0.25925926],
       [0.88333333, 0.11666667],
       [0.52173913, 0.47826087],
       [0.52173913, 0.47826087],
       [0.88333333, 0.11666667],
       [0.52173913, 0.47826087],
       [0.2       , 0.8       ],
       [0.88333333, 0.11666667],
       [0.88333333, 0.11666667],
       [0.74074074, 0.25925926],
       [0.88333333, 0.11666667],
       [0.74074074, 0.25925926],
       [0.52173913, 0.47826087],
       [0.74074074, 0.25925926],
       [0.52173913, 0.47826087],
       [0.74074074, 0.25925926],
       [0.88333333, 0.11666667],
       [1.        , 0.        ],
       [1.        , 0.        ],
       [0.74074074, 0.25925926],
       [0.74074074, 0.25925926],
       [1.        , 0.        ],
       [0.52173913, 0.47826087],
       [0.88333333, 0.11666667],
       [0.74074074, 0.25925926],
       [0.74074074, 0.25925926],
       [0.74074074, 0.25925926],
       [0.88333333, 0.11666667],
       [0.

In [42]:
model_dtc.classes_

array([1, 2], dtype=int64)

In [43]:
y_pred = model_dtc.predict_proba(X_test)[:, 1]
y_pred

array([0.11666667, 0.47826087, 0.25925926, 0.11666667, 0.47826087,
       0.47826087, 0.11666667, 0.47826087, 0.8       , 0.11666667,
       0.11666667, 0.25925926, 0.11666667, 0.25925926, 0.47826087,
       0.25925926, 0.47826087, 0.25925926, 0.11666667, 0.        ,
       0.        , 0.25925926, 0.25925926, 0.        , 0.47826087,
       0.11666667, 0.25925926, 0.25925926, 0.25925926, 0.11666667,
       0.11666667, 0.11666667, 1.        , 0.25925926, 0.11666667,
       0.11666667, 0.11666667, 0.11666667, 0.47826087, 0.        ,
       0.47826087, 0.        , 0.11666667, 0.11666667, 0.11666667,
       0.11666667, 0.11666667, 1.        , 0.11666667, 0.        ,
       0.25925926, 0.11666667, 0.11666667, 0.        , 0.11666667,
       0.11666667, 0.47826087, 0.11666667, 0.        , 0.11666667,
       0.11666667, 0.11666667, 0.11666667, 0.11666667, 0.47826087,
       0.11666667, 0.11666667, 0.11666667, 0.11666667, 0.47826087,
       0.11666667, 0.8       , 0.25925926, 0.47826087, 0.     

In [44]:
roc_auc_score(y_test, y_pred)

0.6248246844319776

Skor dengan model KNN = 0.6297335203366059

Skor dengan model DTC = 0.6248246844319776

Kurang lebih memiliki skor yang sama